### PR TITLE
Tailrunningtask#51

### PR DESCRIPTION
**Pull request by AyubMohamed**
As of right now this command will only tail a task for a task that has completed. There is an issue with tailing live tasks because the `public/logs/live_backing.log` file hasn't been created yet. Maybe we can figure out a way to connect to the worker and get live logs that way but I have been unable to do it so far.

### CODE REVIEW

File: task-log/task_log.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	route := apis.Services["Queue"].BaseURL
+	for _, entry := range apis.Services["Queue"].Entries {
+		if entry.Name == "getLatestArtifact" {
+			route += entry.Route
+			break
+		}
+	}
+
+	route = strings.Replace(route, "<taskId>", taskID, 1)
+	route = strings.Replace(route, "<name>", "public/logs/live_backing.log", 1)
```
Comment (djmitche) [inline task-log/task_log.go]: I think this should use `live.log` instead.  For completed tasks, that seems to work fine in my testing: the `live.log` artifact redirects to `live_backing.log`.

File: task-log/task_log.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+	body, _ := makeGetRequest(route)
+
+	// Check if we got any errors, we are not expecting a json response.
+	var raw map[string]interface{}
+	json.Unmarshal(body, &raw)
+
+	if len(raw) != 0 {
+		// Error, most likely with the taskID
```
Comment (djmitche) [inline task-log/task_log.go]: Good thinking!

File: task-log/task_log.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func makeGetRequest(path string) (b []byte, err error) {
+	resp, err := http.Get(path)
+	if err != nil {
+		panic("Error making request to " + path)
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (djmitche) [inline task-log/task_log.go]: To tail a log "live", I think you'll only need to change this from `ReadAll` to a loop that reads whatever is available form resp.Body, then writes that data to stdout.  That will be a change to the `makeGetRequest` function, since it can no longer return a string.  But that makes sense: if we want to print parts of the task log before the rest of the log exists, we can't return the whole log!

Comment (AyubMohamed) [inline task-log/task_log.go]: The problem is that when we try to tail a "live" log, the get request hangs until the task is completed.

File: task-log/task_log.go (hunk 1 above)
Comment (AyubMohamed) [inline task-log/task_log.go]: For running tasks using `live.log`,  the get request will hang until the task is completed. I chose to use `live_backing.log` for this pr because we don't have a proper way to deal with live tasks at the moment. So in this case using `live_backing.log` will at least give us an error for running tasks.

Comment (jonasfj) [inline task-log/task_log.go]: Yes, `live.log` is a streaming HTTP request, just read it as a stream...
If you're fancy you can do some output buffering before writing to stdout.

File: task-log/task_log.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+func (taskLog) Usage() string {
+	usage := "Usage: taskcluster task-log <taskID>\n"
+	return usage
+}
+
+func (taskLog) Execute(context extpoints.Context) bool {
+	taskID := context.Arguments["<taskID>"].(string)
+
+	// Get route from services.go
+	route := apis.Services["Queue"].BaseURL
```
Comment (jonasfj) [inline task-log/task_log.go]: Yeah, I would not do this...

Client for golang has a signed url method: https://godoc.org/github.com/taskcluster/taskcluster-client-go/queue#Queue.GetLatestArtifact_SignedURL
But since we don't need a signed url here.. I would just hardcode the URL pattern...

**Comment by jonasfj** [comment]
@t0xicCode The tcclient doesn't support streaming responses, and interprets responses as JSON... It's rather domain specific that way.

But yeah, the requests should be done retries with exponential back-off.. I would just look at how that's done in `go-got` (disclaimer I wrote that).

**Comment by t0xicCode** [comment]
@jonasfj I was thinking: tcclient to generate the URL, then using go-got or the standard http client to fetch the data and print it (maybe buffering it). 

Does that sound like a feasible thing? Or is it "overkill"?

**Comment by jonasfj** [comment]
Is there a method to generate URL... I could only find the one to generate
a signed URL...

Den 14. feb. 2017 6.17 PM skrev "Xavier L." <notifications@github.com>:

**Comment by t0xicCode** [comment]
Seems like there isn't. Is it worth it adding such a function to tcclient?

**Comment by jonasfj** [comment]
Probably, maybe @petemoore knows a reason we never added it...

Den 14. feb. 2017 6.24 PM skrev "Xavier L." <notifications@github.com>:

**Comment by djmitche** [comment]
Why not use a signed URL?  It doesn't hurt, and it's something the library already supplies for us..

@AyubMohamed: in the middle of all this back-and-forth, I think the immediate next step is to get streaming working using the URL you've already written the code for.  If we figure out a better way to invent a URL, that should be an easy switch later.

File: task-log/task_log.go (hunk 4 above)
Comment (AyubMohamed) [inline task-log/task_log.go]: That's what I did initially, but i thought we should use that services struct in case the api changes or anything. But i'll just hardcode it for now.

File: apis/doc.go [hunk 5]
```diff
@@ -10,7 +10,7 @@ import (
 )
 
 func init() {
-	for name, service := range services {
+	for name, service := range Services {
```
Comment (djmitche) [inline apis/doc.go]: I think you can revert this change and the change to `services.go` now that you're not using that data in `task_log.go`, right?

File: task-log/task_log.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+	defer resp.Body.Close()
+
+	// Read line by line for live logs.
+	// This will also print the error message for failed requests.
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
+		fmt.Println(scanner.Text())
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
```
Comment (djmitche) [inline task-log/task_log.go]: I'm assuming this always returns an empty string, since you've already read everything in the for loop, above.  So maybe this function should just return an error.  Which means the unmarshalling in Execute isn't necessary.

However, thinking about the error handling here -- I think that Execute would return `true` even for a taskId that doesn't exist, right?  Probably `makeGetRequest` should be looking at the status code in `resp` and signalling an error when the code is not 200 (OK).  In the error case, writing the JSON body to stderr is probably a good approach.

**Comment by AyubMohamed** [comment]
Live logging support now works, I think once we figure out a good way to construct the URL we can create another PR for that issue.

**Comment by petemoore** [comment]
Strange, I thought I added a comment here, but I can't see it now! Anyway, I also think using SignedURL should be fine, and do no harm.

**Comment by jonasfj** [comment]
> Why not use a signed URL? It doesn't hurt, and it's something the library already supplies for us..

Because then you would have to authenticate inorder to view a public log... that's seems excessive.
Many people will likely authenticate with tc-cli, but we shouldn't make unauthenticated APIs require login when it's not needed...

File: task-log/task_log.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	if err != nil {
+		panic("Error making request to " + path)
+	}
+
+	defer resp.Body.Close()
+
+	// Read line by line for live logs.
+	// This will also print the error message for failed requests.
+	scanner := bufio.NewScanner(resp.Body)
+	for scanner.Scan() {
```
Comment (jonasfj) [inline task-log/task_log.go]: Is there any reason to read it line by line?

What if I have a task that plays a cat video as ASCII art using VLC...
Hehe :)

I don't think terminal progress bars like the ones hg or git displays when cloning are necessarily line separated.
Why not just do: `io.Copy(os.Stdout, resp.Body)` ?
If smart you would track length too, match it against content-length if given, and use the length to cut away bytes if retrying the request... So logs don't get duplicated on retries. 


If you want to be really fancy you'll copy from `resp.Body` to an intermediate buffer and flush from that buffer to `os.Stdout` at an interval... Example here:
https://github.com/taskcluster/taskcluster-worker/blob/master/runtime/ioext/copyflush.go
If the intermediate buffer implemented Flush() this CopyAndFlush method could be used.

Oh, and you might want to look into setting stdout into raw mode... Not sure all our logs are actually in raw mode, so maybe we would have to detect that... I think some of the logs uses `\r\n` as line separator which is a good indicator of raw mode.

**Comment by petemoore** [comment]
@jonasfj We should be mindful that one day logs might also be private (and thus signed urls make sense). Maybe we should update the methods that return signed urls to only return signed versions if scopes are actually required, and raw (unsigned) urls if no scopes are required to avoid an additional redirect.

**Comment by djmitche** [comment]
Regarding "fancy", I think line-by-line is good for logging, and this is
indeed a log.  I don't think we need to retry these requests.

**Comment by jonasfj** [comment]
@petemoore 
> Maybe we should update the methods that return signed urls to only return signed versions if scopes are actually required, and raw (unsigned) urls if no scopes are required to avoid an additional redirect.

This can't be determined in the client.

----
To support secret logs, we could add support for options like `--log-path`, that's only specified if it's not `public/live.log`.

Similarly, it would be nice to have an optional `--run-id <runId>` option in case you don't want the latest runId...

**Comment by djmitche** [comment]
OK, so a few follow-on issues to open:
 * generate the URL differently (still an open question on how, or whether, to do this)
 * `--log-path` and `--run-id`
I've added the first point to the agenda for Friday's meeting.

File: task-log/task_log.go (hunk 7 above)
Comment (AyubMohamed) [inline task-log/task_log.go]: The problem with io.Copy is that it copies from src to dst until EOF is reached. For live logging the request doesn't come back until the log is complete, so there is no EOF at the moment.

Comment (djmitche) [inline task-log/task_log.go]: I think io.Copy would probably work, but it might return chunks of log that aren't newline-aligned, which is sort of ugly.

I think for logs, which are line-oriented, the line-by-line approach is right.

Comment (AyubMohamed) [inline task-log/task_log.go]: Actually my mistake, this should work. I just don't know how big the chunk or how often it copies to `os.Stdout`.
